src/module.c

```diff
@@ -3540,6 +3540,241 @@ int RM_PublishMessageShard(RedisModuleCtx *ctx, RedisModuleString *channel, Redi
+        listSetFreeMethod(moduleChannelSubscriber->subscriptions, decrRefCountVoid);
+        dictAdd(moduleChannelSubscribers, ctx->module, moduleChannelSubscriber);
+    } else {
+        moduleChannelSubscriber = dictGetVal(de);
+    }
+
+    if (listSearchKey(moduleChannelSubscriber->subscriptions, channel)) {
+        return;
+    }
+
+    /* Add the channel to the list of subscriptions. */
+    robj *channelToAdd = dupStringObject(channel);
+    listAddNodeTail(moduleChannelSubscriber->subscriptions, channelToAdd);
+
+    /* Add the module to the (channel -> list of module subscribers hash table). */
+    list *subscribers = NULL;
+    de = dictFind(channelModuleSubscribers, channelToAdd);
+    if (de == NULL) {
+        subscribers = listCreate();
+        dictAdd(channelModuleSubscribers, channelToAdd, subscribers);
+        incrRefCount(channelToAdd);
+    } else {
+        subscribers = dictGetVal(de);
+    }
+    listAddNodeTail(subscribers, ctx->module);
+
+    serverLog(LL_DEBUG,"Module %s subscribed to channel %s.", ctx->module->name, (char *)channelToAdd->ptr);
+
+    return;
+}
+
+/**
+ * This API unsubscribe(s) the module from the given channel.
+ * 
+ */
+void RM_UnsubscribeFromChannel(RedisModuleCtx *ctx, RedisModuleString *channel) {
+    dictEntry *de = dictFind(moduleChannelSubscribers, ctx->module);
+    if (de == NULL) {
+        return;
+    }
+
+    RedisModuleChannelSubscriber *moduleChannelSubscriber = dictGetVal(de);
+
+    /* No subscription found for module. */
+    if (listLength(moduleChannelSubscriber->subscriptions) == 0) {
+        return;
+    }
+
+    listNode *ln = listSearchKey(moduleChannelSubscriber->subscriptions, channel);
+
+    if (ln == NULL) {
+        return;
+    }
+
+    /* Find and remove the module subscriber to the channel. */
+    removeModuleFromChannelSubscription(ctx->module, ln->value);
+    listDelNode(moduleChannelSubscriber->subscriptions, ln);
+    if (listLength(moduleChannelSubscriber->subscriptions) == 0) {
+        cleanupModuleChannelSubscriber(ctx->module, moduleChannelSubscriber);
+    }
+    serverLog(LL_DEBUG,"Module %s unsubscribed from channel %s.", ctx->module->name, (char *)channel->ptr);
+}
+
+/**
+ * List all global channels a module is subscribed to.
+ * 
+ * Returns an array of RedisModuleString pointer in which the last entry is NULL. 
+ * The array must be freed by the caller via `RM_FreeGlobalChannelSubscriptionList`.
+ */
+RedisModuleString **RM_GlobalChannelSubscriptionList(RedisModuleCtx *ctx, size_t *numchannels) {
+    *numchannels = 0;
+    dictEntry *de = dictFind(moduleChannelSubscribers, ctx->module);
+    if (de == NULL) {
+        return NULL;
+    }
+
+    RedisModuleChannelSubscriber *moduleChannelSubscriber = dictGetVal(de);
+    listIter li;
+    listRewind(moduleChannelSubscriber->subscriptions, &li);
+
+    int channelCount = listLength(moduleChannelSubscriber->subscriptions);
```

Comment:
I am not sure this is better, but by returning a list with a dedicated free method caller can simply free the list and no need for RM_FreeGlobalChannelSubscriptionList. there might be some performance related impact in case this list is taken many times...

Reply:
I'm not sure how all the languages used for writing module would deal with this.

I looked for similar pattern in the code and realized there was similar implementation for `RM_GetClusterNodesList` via `RM_FreeClusterNodesList` https://github.com/redis/redis/blob/unstable/src/module.c#L8822